Kitodo-DataAccess/src/main/java/org/kitodo/dataaccess/format/xml/XMLWriter.java

```diff
@@ -0,0 +1,352 @@
+import java.io.OutputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.NoSuchElementException;
+import java.util.Optional;
+import java.util.TreeMap;
+
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.transform.OutputKeys;
+import javax.xml.transform.Transformer;
+import javax.xml.transform.TransformerException;
+import javax.xml.transform.TransformerFactory;
+import javax.xml.transform.dom.DOMSource;
+import javax.xml.transform.stream.StreamResult;
+
+import org.kitodo.dataaccess.AccessibleObject;
+import org.kitodo.dataaccess.IdentifiableNode;
+import org.kitodo.dataaccess.LangString;
+import org.kitodo.dataaccess.Literal;
+import org.kitodo.dataaccess.Node;
+import org.kitodo.dataaccess.NodeReference;
+import org.kitodo.dataaccess.ObjectType;
+import org.kitodo.dataaccess.RDF;
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+
+/**
+ * Transforms a linked data structure to an XML file.
+ *
+ * @author Matthias Ronge
+ */
+public class XMLWriter {
+    /**
+     * String to pass to the transformer to switch on indenting.
+     */
+    private static final String TRANSFORMER_INDENT_TRUE = "yes";
+
+    /**
+     * String to pass to the transformer to set the indenting width in
+     * characters.
+     */
+    private static final String TRANSFORMER_INDENT_VALUE = "{http://xml.apache.org/xslt}indent-amount";
+
+    /**
+     * Returns the type of the node, or {@code rdf:Description} if not set.
+     *
+     * @param node
+     *            node whose type shall be returned
+     * @return the type of the node, or {@code rdf:Description}
+     */
+    private static String getTypeOrDefault(AccessibleObject node) {
+        try {
+            return node.getType();
+        } catch (NoSuchElementException e) {
+            return RDF.DESCRIPTION;
+        }
+    }
+
+    /**
+     * Populates a DOM element with the data from this node.
+     *
+     * @param node
+     *            node to transform
+     * @param xml
+     *            the dom document, needed to create the children
+     * @param abbr
+     *            the namespace handler initialized at the root node
+     * @param out
+     *            element to populate
+     * @throws NoSuchElementException
+     *             if a relation links to an empty set of objects. (If so, the
+     *             object model is corrupt.)
+     */
+    private static void nodeToElement(Node node, Document xml, Namespaces abbr, Element out) {
+        TreeMap<Long, ObjectType> orderedElements = new TreeMap<>();
+        HashMap<String, Collection<ObjectType>> bagElements = new HashMap<>();
+        long first = Long.MAX_VALUE;
+        long last = 0;
+
+        if ((node instanceof Node) && (node instanceof IdentifiableNode)) {
```

Comment:
Condition check _node instanceof Node_ should be removed, and if any check is needed here, it is null check _(Objects.nonNull(node))_.